### ISSUE TITLE
Segmentation fault at random intervals when extracting pictures

### ISSUE THREAD

**Issue by jameswyper** [robinst/taglib-ruby#91]
I'm attempting to use taglib-ruby to extract metadata from my large collection of flac files and it's segfaulting after reading "some" of them (where "some" varies randomly across runs between 100 and 6000+).
I've reproduced the issue with a single file - I expect this will fail for any flac with embedded art (but I can send you the offending flac if you want).

```
require 'taglib'

def trigger(file)
    TagLib::FLAC::File.open(file) do |f|
        f.picture_list.each do |p|
            x = p.data
        end
    end
end

10000.times do |i|
    trigger("/home/james/test.orig.flac")
    if (i % 100 == 0) then puts "iteration #{i}" end
end
```
it will also fail if you replace "x = p.data" with just "p.data" but consistently run to completion if you comment that line out entirely (in other words don't attempt to access the picture data at all).
I'm getting this on ruby 2.5.1p57 and 2.5.5p257 on Ubuntu 18.04/19.04.  The gem is up to date.

I've not got much experience debugging at the segfault level but if you think I can provide any further useful info please ask.  Thanks.

**Comment by jameswyper** [comment]
I've just tested with a flac file from Archive.org and had the same problem - hopefully that makes it easier to reproduce.  The first time I ran with this file all 10,000 iterations completed, but subsequent test runs failed partway through.

https://archive.org/download/78_you-are-my-sunshine_paul-ric...sticker-rice-brothers-gang_gbia0000125a/You%20Are%20My%20Sunshine%20-%20Paul%20Ric...sticker.flac

My 2.5.5 Ruby was running on a Ubuntu 19.04 VM inside Virtualbox (the 2.5.1 runs on bare metal).

I also changed my test code a little:

```
require 'taglib'

def trigger(file)
    TagLib::FLAC::File.open(file) do |f|
        #f.picture_list.each do |p|
            #p.data
        #end
        f.xiph_comment.field_list_map.each do |tag,value|
            tag
            value
        end
    end
end

100000.times do |i|
    trigger("/home/james/sunshine.flac")
    if (i % 100 == 0) then puts "iteration #{i}" end
end
```
(note 10x more iterations this time)
The code above has run OK (twice) but if I uncomment the f.picture_list... block it will fail.  It appears to fail faster if I also uncomment the p.data line inside the block.


I've had a look at the code and - bearing in mind I really don't understand SWIG - it looks as if the code that's present to free / unlink variables / objects for Vorbis comments etc when the FLAC file is closed might not be there for the picture objects..

Hope this helps.

**Comment by jacobvosmaer** [comment]
>it looks as if the code that's present to free / unlink variables / objects for Vorbis comments etc when the FLAC file is closed might not be there for the picture objects.

That does sound suspicious. The picture list appears to be part of the "xiph comment" so you'd expect objects to get unlinked here: https://github.com/robinst/taglib-ruby/blob/86567b7b689cdf276264552a5dd10476ef5c33cc/ext/taglib_flac/taglib_flac.i#L78-L81

**Comment by jameswyper** [comment]
I've forked and made some changes here https://github.com/jameswyper/taglib-ruby/tree/issue-91-fix-memory-leak-flac-picture
It would be great if someone could review them as it's my first time doing this sort of thing.  In summary I have:

1. Copied the flac.flac file in test/data to flac_nopic.flac and run `metaflac --remove --block-type=PICTURE flac_nopic.flac` to get rid of the picture
2.  Added new test flac_picture_memory_test - this runs tests on both flac files (I wanted to be sure that my changes didn't break anything if a picture wasn't embedded).  This new test fails on release 1.0.0
3. Changed taglib_flac.i to unlink / remove tracking for each entry in the picture list (at least, that's what I hope I've done)
4. (I had also messed around with flac_create.cpp but was getting nowhere so reverted my changes)
5. Built and run a successful rake test

If everything looks OK I'll raise a pull request.  Thanks.

**Comment by jacobvosmaer** [comment]
It's usually easier to just open a PR and make clear it's not for merging yet. Now we have to click around just to see what changed on your branch.

I think it's this: https://github.com/robinst/taglib-ruby/compare/master...jameswyper:issue-91-fix-memory-leak-flac-picture

**Comment by jacobvosmaer** [comment]
It looks reasonable. I don't see why this can't be submitted as a PR.

**Comment by robinst** [comment]
Fix released in 1.0.1!: https://github.com/robinst/taglib-ruby/blob/master/CHANGELOG.md#101---2020-03-25
